[RFC][DNM] cmake: RFC for single libraries to use CMake dependencies and remove whole-archive

tejlmand opened the pull request:
This RFC present a possibility on how subsys Bluetooth and subsys net can be changed into a single library while still use the zephyr KConfig principle as today.
See issue: #8441 

This principle can further be extended to other libraries in Zephyr.

The RFC remove the need to use --start-group /--end-group as well as the --whole-archive
The changes made in this RFC remove the need for whole archive on subsys_net and subsys_bluetooth as seen in the output below, where those libraries are outside the whole-archive.
The whole arhive issue is also mentioned in: #6961 

Furthermore net and bluetooth are removed from ${ZEPHYR_LIBS_PROPERTY} as described in #8441

This also ensures that CMake normal dependency handling is obtained.

Furthermore, specific compile flags on sources, such as those found in 
subsys/bluetooth/controller/bluetooth_controller.cmake (was: subsys/bluetooth/controller/CMakeLists.txt) are now done using, 
  SET_SOURCE_FILES_PROPERTIES( ${PRIVATE_SOURCES} PROPERTIES COMPILE_FLAGS -Ofast), which has two advantages over current form:
1) Compile clags as property on files automatically has precendence over library link flags, thus no special handling is needed as opposed to: PR #5110
where the order of libraries is very important
2) Compile flags only added for source files are not accidentally propagate, #8438 

The principle used when including *.cmake files are further described here:
https://crascit.com/2016/01/31/enhanced-source-file-handling-with-target_sources/

Another nice part of the proposed structured, is that only the entry point to the library contains a CMakeLists.txt file, i.e. subsys/net/CMakeLists.txt and subsys/bluetooth/CMakeLists.txt 
and other files are given more descriptive names,making navigation easier when having multiple *.cmake files open, instead of all files being named CMakeLists.txt

Not addressed in this RFC / commit:
Zephyr has an extended version of add_subdirectory called add_subdirectory_ifdef.
In this RFC, the straight forward way of using
if(KCONFIG_FLAG)
  include(...)
endif()
but I see no problem in creating an extended version, called include_ifdef(KCONFIG_FLAG ....) if desired.

--whole-archive has not been completely removed, as there are still libraries left which needs tthis flag, but I hope the principle used is enough for discussion and then rest of libraries can be addressed in future commit.


Output from Zephyr when linking sample app bluetooth/peripheral_hr:
[139/144] : && ccache /opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/bin/arm-zephyr-eabi/arm-zephyr-eabi-gcc    zephyr/CMakeFiles/zephyr_prebuilt.dir/misc/empty_file.c.obj  -o zephyr/zephyr_prebuilt.elf  -T zephyr/linker.cmd -Wl,-Map=/projects/github/tejlmand/zephyr/samples/bluetooth/peripheral_hr/build/zephyr/zephyr.map -u_OffsetAbsSyms -u_ConfigAbsSyms -e__start -Wl,--whole-archive libapp.a zephyr/libzephyr.a zephyr/arch/arm/core/libarch__arm__core.a zephyr/arch/arm/core/cortex_m/libarch__arm__core__cortex_m.a zephyr/arch/arm/core/cortex_m/mpu/libarch__arm__core__cortex_m__mpu.a zephyr/lib/libc/minimal/liblib__libc__minimal.a zephyr/drivers/gpio/libdrivers__gpio.a zephyr/drivers/serial/libdrivers__serial.a zephyr/drivers/entropy/libdrivers__entropy.a -Wl,--no-whole-archive zephyr/kernel/libkernel.a zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj -L"/opt/zephyr-sdk/sysroots/armv5-zephyr-eabi/usr/lib/arm-zephyr-eabi/6.2.0/armv7e-m" -L/projects/github/tejlmand/zephyr/samples/bluetooth/peripheral_hr/build/zephyr -lgcc -Wl,--print-memory-usage zephyr/subsys/bluetooth/libsubsys_bluetooth.a -u_hci_driver_init zephyr/subsys/net/libsubsys_net.a zephyr/libzephyr.a zephyr/kernel/libkernel.a zephyr/liboffsets.a -nostartfiles -nodefaultlibs -nostdlib -static -no-pie -Wl,-X -Wl,-N -Wl,--gc-sections -Wl,--build-id=none && :

EDIT: This is part of umbrella issue: #8827

tejlmand committed:
cmake: Created a function which creates a list based on IFDEF / IFNDEF.

RFC for #8439

- This commits adds a new function for creating and populating a list
  with IFDEF and IFNDEF as well as specifying source to ensure default
  path pre-fixing.

- function is: zephyr_list

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Moving of TOOLCHAIN_LIBS and _ConfigAbsSyms settings

Fixes: #8440

This commit move the link flag ${TOOLCHAIN_LIBS} from the global list to
zephyr_interface and -u_ConfigAbsSyms to the zephyr library in order to
give a better structure as described in #8440

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Zephyr interface depends upon kernel

Fixes: #8440

This commit moves kernel from the list in <zephyr_base>/CMakeLists.txt
and place it as a link dependency when linking to zephyr_interface.

It also links kernel to offsets to ensure correct link dependencies.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Example showing the use of fewer libraries.

RFC for #8441

The RFC creates fewer libraries and uses CMake link libraries commands
to utilize CMake dependency handling
- subsys bluetooth has been converted into a single library
- subsys network has been converted into a single library

It uses the populate source list presented in: #8442

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Offset library linking flags

Fixes: #8440

This commit moves the link flag  -u_OffsetAbsSyms which is required
when linking to offset static library.

The library offset thus propagates the flags correctly during linking.
It also ensures knowledge about link flags is present where it
correctly belongs.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand commented:
Note, the main part of this RFC is the reduced numbers of libs which then leads to the fact that it will be easier to remove the -Wl,--whole-archive.
But the fact that a library really becomes a library and not just a collection of the files in a subfolder gives a much nicer architectural structure.
This gives a better correspondence between a library and a subsystem in Zephyr context.
Which again is an advantage when understanding the structure in Zephyr.

This especially become apparent in an IDE, an example from Eclipse can be seen below.
The screen shots show how the numbers of libs marked in read is reduced by this RFC.
The libs marked in blue are other candidates to merge to single libs.
![example_current](https://user-images.githubusercontent.com/9802583/41658453-aed707c8-7496-11e8-842e-651454e5c6db.png)
and after the change:
![example_single_lib](https://user-images.githubusercontent.com/9802583/41658469-b962dbfe-7496-11e8-9331-39332259e8f2.png)

carlescufi commented:
@tejlmand can you rebase please?
@SebastianBoe any thoughts on this one?

SebastianBoe commented:
> Compile flags only added for source files are not accidentally propagate, #8438 

This point can be removed since #8438 has been resolved (the fix is blocked by awaiting PR review).

SebastianBoe commented on cmake/app/boilerplate.cmake:
```diff
@@ -313,3 +313,6 @@ foreach(boilerplate_lib ${ZEPHYR_INTERFACE_LIBS_PROPERTY})
 endforeach()
+
+target_link_libraries_ifdef(CONFIG_BT      app subsys_bluetooth)
+target_link_libraries_ifdef(CONFIG_NET_BUF app subsys_net)
```
Core build script code such as boilerplate.cmake must be agnostic to the details of subsystems.

For this use-case, one should take advantage of the CONFIG_APP_LINK_WITH_${boilerplate_lib_upper_case}
feature. See above foreach loop.

tejlmand replied:
yes, as I stated in the commit message:

> This is currently a temporary solution to ensure all samples and tests
> builds correctly.
> 
> As part of the RFC it should be decided what the optimal approach for
> linking should be.
> 

So the loop is a candidate.
Another approach could be to target_link_libraries where the subsystem is actually included, e.g.
as part of:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/CMakeLists.txt#L1-15

So when a subsystem is included, it is also linked to app.

SebastianBoe commented on subsys/bluetooth/CMakeLists.txt:
```diff
@@ -1,10 +1,20 @@
+# easy understanding og what is available Mesh is part of host in this example.
 
-target_include_directories(subsys__bluetooth INTERFACE ${CMAKE_CURRENT_SOURCE_DIR})
+add_library(subsys_bluetooth STATIC "")
```
This breaks userspace/kernelspace memory separation.

To protect kernelspace and userspace memory regions differently the linker needs to place them in different sections.

Today this is implemented by having CMakeLists.txt iterate over the libraries in ZEPHYR_LIBS_PROPERTY and guess based on their names which one's should be in kernelspace and which ones should be in userspace. Not great, I know.

It is implemented here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/CMakeLists.txt#L526

This breaks when you take subsys_bluetooth out of the global list of Zephyr CMake Libraries.

To do this re-organization we need to first find some other mechanism to split up the kernelspace and userspace memory regions.

tejlmand replied:
Yes, this I wasn't aware about, that the list were used for placing libs different mem location.
See further comment below, where I list all places ZEPHYR_LIBS_PROPERTY is used

SebastianBoe commented on subsys/bluetooth/CMakeLists.txt:
```diff
@@ -1,10 +1,20 @@
-add_subdirectory_ifdef(CONFIG_BT_HCI host)
-add_subdirectory_ifdef(CONFIG_BT_CTLR controller)
-add_subdirectory_ifdef(CONFIG_BT_SHELL shell)
+include(${CMAKE_CURRENT_LIST_DIR}/common/bluetooth_common.cmake)
```
This also breaks every other use-case that is covered by the existence of ZEPHYR_LIBS_PROPERTY.

So what is lacking from this RFC is an overview of all the cases where a global list of libraries is used and new strategies for each of them.

tejlmand replied:
ZEPHYR_LIBS_PROPERTY is only used in <zephyr-repo>/CMakeLists.txt file.

And it is used for three things, as I can see:
1) Looping the libs to ensure they are not empty:
https://github.com/zephyrproject-rtos/zephyr/blob/master/CMakeLists.txt#L499-L508
and if the libray is imported, then it is disregarded.
Thus, only libraries created by Zephyr's build system is checked.
In CMake, if you specify a libray and don't add sources, then you get the following error:
'No SOURCES given to target: <library>'
(except for INTERFACE libs, but those are not checked here)
So this RFC should also remove the need for this checked.
Not sure why the check is here, but could imagine that using a list with all libraries, is the root cause here, as someone could add invalid targets to the list
2) at set(zephyr_lnk ...)
https://github.com/zephyrproject-rtos/zephyr/blob/master/CMakeLists.txt#L706-L713
Which will be obsolete with this RFC
3) Kernel / userspace
https://github.com/zephyrproject-rtos/zephyr/blob/master/CMakeLists.txt#L526
This is a real problem.
Will look into a proposal, and update the RFC with that.

SebastianBoe replied:
About the empty check. It is OK to drop this check, it was added out of convenience since the list already existed.

The CMake version that I was using at the time was giving an obscure error message about the wrong language of source files being used and nothing about the library being empty, which is why the check was introduced.

tejlmand replied:
I have added a change with improved handling of kernel / app memory space location.

A custom target memory_space_kernel is added.
The memory_space_kernel is then populated with those targets, such as kernel, subsys_bluetooth, subsys_net, etc.

Having much fewer libraries makes it less cumbersome to specify which libraries should be in kernel space explicitly, and much clearer to the developer.
Furthermore it gives developers using Zephyr to place their own libs in kernel space.

A general function has been introduced, which follows same style as CMake set_property but handles specific Zephyr properties, currently MEMORY_SPACE kernel, but may be extended in future.
Example:
zephyr_set_property(TARGET subsys_net MEMORY_SPACE kernel)

Advantage of this approach over current:
1) Libs are not place in kernel space based on guessing from their location / name
2) Libs are explicitly stated when they are kernel space candidate
3) Using of target (over global variable) ensures that spelling mistakes are causing an error, eg. writing  zephyr_set_property(TARGET subsys_nett MEMORY_SPACE kernel) results in:   get_target_property() called with non-existent target "subsys_nett".
Call Stack (most recent call first):
  ../../../subsys/net/CMakeLists.txt:25 (zephyr_set_property)
4) Using target and target properties that supports generator expressions removes the configure time dependency of CMake code ordering. i.e zephyr_set_property() can be used in anyplace of cmake code, whereas previously any calls that populating ${ZEPHYR_LIBS_PROPERTY} had to be placed before: https://github.com/zephyrproject-rtos/zephyr/blob/8a74e7056dc6e405b1336e047957d143af5b72d9/CMakeLists.txt#L534-L625


Disadvantage:
1) Having to write zephyr_set_property(TARGET subsys_net MEMORY_SPACE kernel). However, with the reduced number of libraries that this RFC propose, this becomes a much smaller issue, and I think the pros out weights the cons.

tejlmand commented:
> >   Compile flags only added for source files are not accidentally propagate, #8438

> This point can be removed since #8438 has been resolved (the fix is blocked by awaiting PR review).

No, cause if it is approved to limit the number of libs this way, it is important to be aware that this RFC still fixes issue #8438

SebastianBoe commented on subsys/bluetooth/controller/bluetooth_controller.cmake:
```diff
@@ -0,0 +1,60 @@
+set_target_properties(subsys_bluetooth PROPERTIES EXCLUDE_FROM_ALL False)
```
Why do we want to include subsys_bluetooth in all?

tejlmand replied:
You're not including it in all, instead you're not excluding it.
but this line should not be here.

and i'm actually a bit puzzled how it got in here so I will remove as well as the line in bluetooth_common.cmake.

SebastianBoe commented on subsys/bluetooth/controller/bluetooth_controller.cmake:
```diff
@@ -0,0 +1,60 @@
+
+# We must tell the linker not to remove the _hci_driver_init function during linking.
+# Notice: This is done in the same file where we have the knowledge that hci_driver.c is added to
+#         the library target.
```
The link command must be kept as clean as possible.

Perhaps we could achieve this by modifying SYS_INIT and/or the linker scripts to force symbols into the build.

tejlmand replied:
You need someone OUTSIDE the library to use the symbol.
Just placing an internal reference won't help you.

Placing it as a link dependency actually makes it very clear, that this is needed when using this module.
And the flag is activated based on the KConfig setting, which includes the file requiring the flag.
In this case: hci_driver.c and -u_hci_driver_init are only added to the library when CONFIG_BT_CTRL is enable in KConfig.
This makes things that belongs together also being placed together.

SebastianBoe replied:
> Placing it as a link dependency actually makes it very clear, that this is needed when using this module.

I agree that it reads cleanly that this flag is needed when using this module. What I want is for this flag to not be needed at all.

Inuitively, users should as often as possible be able to work with a C-based interface, like the SYS_INIT macro. And then the build will behave as they expect it to without them having to also keep in sync metadata in some other file, like this build script.

It is redundant to invoke SYS_INIT and also to say that this symbol is undefined. This redundancy leads to user errors when these two pieces of information get out-of-sync. This redundancy also hurts the learning curve as the user must learn that this is necessary through documentation.

SebastianBoe replied:
It seems we will not have to fake an unresolved reference for all drivers.

E.g. bt_hci_driver can be resolved like this:

https://github.com/SebastianBoe/zephyr/commit/d8d6fd7835a9715fceec7455e397e2894a912d97

This has the added benefit of cutting down a bit on runtime overhead and flash size.

Perhaps if it is rare that the user needs to insert this reference this is acceptable ...

I can't think of any simple mechanisms that would avoid this metadata (although several complex ones could ...).

@jhedberg has confirmed that it is OK to drop (it doesn't actually exist) runtime-support for multiple driver implementations. So the above patch at least should be safe ...

tejlmand replied:
So your suggestion is to remove of a struct containing the function, with directly called functions.

It's good that @jhedberg has confirmed that this is ok in this particular case, but will this be a solution that can work everywhere ?

The reason I ask is because having to write code in different way seems like limiting the features of C in Zephyr.

Currently, I have seen the problem with the _hci_driver_init in relation to the SYS_INIT macro.
I currently don't know how many places the SYS_INIT is used in a way that causes same problem as with the _hci_driver_init, but SYS_INIT is used 191 times i C code.
And some of those might not be able to change this way.

However, it will be safe to add -u<function> to those that benefits from it, such as _hci_driver_init.

Get me right on this: Simplifying code when it can be done, while at the same time remove this problem is good.
But we should be careful not to enforce people to write drivers in a specific way or limit the features of C.

tejlmand replied:
@SebastianBoe Please pay attention to the comment provided by @andrewboie 

> Have you checked the effect on various footprint sizes when modifying the archive settings? Are there any differences? I mention this because we use these settings to save space in the kernel. For example, if the application does not use any of the workqueue APIs, then the code in kernel/system_work_q.c, instead of pulling the rest of the workqueue code in, gets dropped too and the k_sys_work_q_init function never gets run at boot.

SYS_INIT is not always the same as the symbol should be considered undefined by linker.
In some cases (kernel), the code should be dropped

tejlmand replied:
The commit: ffa3a2b introduces a property you can set on the target.

This will search the target for the devconfig.sysinit section after creation and any init symbols will be populated into the linker.

This removes the need for knowing/specifying the exact symbol names, thus:
`target_link_libraries(<target> INTERFACE -u<symbol> -u<...> ...)`
 is no longer needed.

Instead, a property is set on the target when it is known to contain initialization symbols:
`zephyr_set_property(TARGET <target> SYSINIT)`
eg.
`zephyr_set_property(TARGET subsys_bluetooth SYSINIT)`

SebastianBoe commented on subsys/bluetooth/controller/bluetooth_controller.cmake:
```diff
@@ -0,0 +1,60 @@
+# Also it prevents any unintended propagation of flags as described in: #<>
+if(CONFIG_BT_CTLR_FAST_ENC)
+  SET_SOURCE_FILES_PROPERTIES( ${PRIVATE_SOURCES} PROPERTIES COMPILE_FLAGS -Ofast)
+endif()
```
To limit complexity we need to use few and orthogonal concepts to achieve as many use-cases as possible.

In this case, it means that we are already using a mechanism to partition flags between sets of source files. libraries. So we should continue using it.

There exists no mapping between 'CMake libraries' and the 'modules' in zephyr/subsys/, zephyr/lib, zephyr/drivers/, zephyr/ext, etc. and introducing such a mapping would be artificial and harmful (because we can no longer use target interface properties to organize code within a subsys, lib, ext component etc.).

AFAICT today this is more important than keeping IDE project windows clean. I have not looked deeply into the problems that occur from "too many libraries", I hope some of them could be mitigated through other means than reducing the number of libraries, such as using library properties that the IDE is aware of.

SebastianBoe commented:
I see that this RFC proposes to replace zephyr_library_sources with zephyr_populate_list.

Is that the extent of the API that would be removed, or is the proposal to also remove the zephyr_* and zephyr_library_* API?

The reason I ask is that this is a very big change, so it would require a very big motivation. Also, I like the proposal to remove --whole-archive by leveraging CMake libraries more and deprecating the global list of libraries. This is a worthy goal, but it is not obvious to me whether this depends on or is independent frommigrating off of the zephyr_* and zephyr_library_* API's.

tejlmand commented:
> I see that this RFC proposes to replace zephyr_library_sources with zephyr_populate_list.

> Is that the extent of the API that would be removed, or is the proposal to also remove the zephyr_* and zephyr_library_* API?

> The reason I ask is that this is a very big change, so it would require a very big motivation. Also, I like the proposal to remove --whole-archive by leveraging CMake libraries more and deprecating the global list of libraries. This is a worthy goal, but it is not obvious to me whether this depends on or is independent from migrating off of the zephyr_* and zephyr_library_* API's.

Firstly, for this RFC, i needed the possibility to add sources without breaking the part of the existing CMake libraires that used existing API, but there are also other reasons:
1) This RFC cannot use the target sources as-is, primarily due to the fact the ${CMAKE_CURRENT_SOURCE_DIR} is used
2) As #8439 also states, I believe it is better to use plain CMake, instead of having people referring to multiple sets of API's, especially as the Zephyr API simply states: 
> # The methods are wrappers around the CMake target_* functions. See                                                                                           
> # https://cmake.org/cmake/help/latest/manual/cmake-commands.7.html for                                                                                        
> # documentation on the underlying target_* functions. 
Then let's use the underlying functions

The part of needing to populate the ZEPHYR_LIBS_PROPERTY when creating new libraries also disappears, so basically, a lot of this removes the need for having all those wrapper functions.

But I guess that discussion is more suited in #8439

tejlmand commented:
> The reason I ask is that this is a very big change, so it would require a very big motivation. Also, I like the proposal to remove --whole-archive by leveraging CMake libraries more and deprecating the global list of libraries. This is a worthy goal, but it is not obvious to me whether this depends on or is independent from migrating off of the zephyr_* and zephyr_library_* API's.

Yes, that is a large change, but I would say it's worth the effort, and I would happily do it, if this proposal is accepted, as I believe it contains several benefits in the overall architecture of the build system.

tejlmand committed:
cmake: Kernel and app memory space update

Fixes: #8851 #8441

This commit updates the handling of memory location for libraries in
Zephyr during linking.

This is needed in order to be able to add libraries after
ZEPHYR_LIBS_PROPERTY has been used in CMake.

It introduces a generic zephyr_set_property function which can be
extended in future.

At start is supports the property:
- MEMORY_SPACE kernel
which can be used to control memory space location for libraries when
KConfig setting CONFIG_APPLICATION_MEMORY is enabled.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand commented:
I have added a change with improved handling of kernel / app memory space location.

A custom target memory_space_kernel is added.
The memory_space_kernel is then populated with those targets, such as kernel, subsys_bluetooth, subsys_net, etc.

Having much fewer libraries makes it less cumbersome to specify which libraries should be in kernel space explicitly, and much clearer to the developer.
Furthermore it gives developers using Zephyr to place their own libs in kernel space.

A general function has been introduced, which follows same style as CMake set_property but handles specific Zephyr properties, currently MEMORY_SPACE kernel, but may be extended in future.
Example:
zephyr_set_property(TARGET subsys_net MEMORY_SPACE kernel)

Advantage of this approach over current:
1) Reduced numbers of libraries significantly reduces the risk of hitting the 32 limit as described in: #7703
2) Libs are not place in kernel space based on guessing from their location / name
3) Libs are explicitly stated when they are kernel space candidate
4) Using of target (over global variable) ensures that spelling mistakes are causing an error, eg. writing  zephyr_set_property(TARGET subsys_nett MEMORY_SPACE kernel) results in:   get_target_property() called with non-existent target "subsys_nett".
Call Stack (most recent call first):
  ../../../subsys/net/CMakeLists.txt:25 (zephyr_set_property)
5) Using target and target properties that supports generator expressions removes the configure time dependency of CMake code ordering. i.e zephyr_set_property() can be used in anyplace of cmake code, whereas previously any calls that populating ${ZEPHYR_LIBS_PROPERTY} had to be placed before: https://github.com/zephyrproject-rtos/zephyr/blob/8a74e7056dc6e405b1336e047957d143af5b72d9/CMakeLists.txt#L534-L625



Disadvantage:
1) Having to write zephyr_set_property(TARGET subsys_net MEMORY_SPACE kernel). However, with the reduced number of libraries that this RFC propose, this becomes a much smaller issue, and I think the pros out weights the cons.

andrewboie commented:
Have you checked the effect on various footprint sizes when modifying the archive settings? Are there any differences?

tejlmand commented:
> Have you checked the effect on various footprint sizes when modifying the archive settings? Are there any differences? 

Currently the kernel library behave the same. So no changes here.
The kernel lib is already outside the --whole-archive flag.

However, as discussed here: https://github.com/zephyrproject-rtos/zephyr/pull/8451/#discussion_r198883689

there need to be an agreement for those libraries which today are whole-archived.
Such as the Bluetooth, which contains _hci_driver_init that must not be removed by linker.
(and also uses the SYS_INIT macro)

Your comment makes it clear, that we cannot just relate all SYS_INIT macro call to the undefined flag, as @SebastianBoe mentions in the comment.
In the kernel case, SYS_INIT is used, but linker should still be able to drop the table.

We must either specify exactly what symbols should be kept, as proposed using the -u_hci_init_driver.
This can easily be done for all libs the are currently whole-archived today, as one can search for the SYS_INIT.

Or explicitly state that this libs should keep symbols used with SYS_INIT (and DEVICE_AND_API_INIT).
So one could specify a property on subsys_bluetooth, e.g:
zephyr_property(TARGET subsys_bluetooth SYS_INIT)
such approach would minimize the work of finding all SYS_INIT.

So that without the property the linker can drop code, such as in the kernel lib.



That to

tejlmand committed:
cmake: Merged n times arch_arm_Y libs into soc_arm and arch_arm

Fixes: #8441

This commit merges the multiple sub libraries under soc_arm and
arch_arm, e.g. arch_arm_Y_Z into a single arch_arm libraries.

It also moves arch related stray files from libzephyr.a and places them
inside the arch_${ARCH} and soc_${ARCH} library, see #8826.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Merged n times arch_arc_Y libs into soc_arc and arch_arc

Fixes: #8441

This commit merges the multiple sub libraries under soc_arc and
arch_arc, e.g.  arch_arc_Y_Z into a single arch_arc libraries.

It also moves arch related stray files from libzephyr.a and places them
inside the soc_${ARCH} and arch_${ARCH} library, see #8826.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Merged n times arch_nios2_Y libs into soc_nios2 and arch_nios2

Fixes: #8441

This commit merges the multiple sub libraries under soc_nios2 and
arch_nios2, e.g. arch_nios2_Y_Z into a single arch_nios2 libraries.

It also moves arch related stray files from libzephyr.a and places them
inside the arch_${ARCH} library, see #8826.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Merged n times arch_posix_Y libs into soc_posix and arch_posix

Fixes: #8441

This commit merges the multiple sub libraries under soc_posix and
arch_posix, e.g. arch_posix_Y_Z into a single arch_posix libraries.

It also moves arch related stray files from libzephyr.a and places them
inside the arch_${ARCH} library, see #8826.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Merged n times arch_riscv32_Y libs into soc_- / arch_riscv32

Fixes: #8441

This commit merges the multiple sub libraries under soc_riscv32 and
arch_riscv32, e.g. arch_riscv32_Y_Z into a single arch_riscv32
libraries.

It also moves arch related stray files from libzephyr.a and places them
inside the arch_${ARCH} library, see #8826.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Merged n times arch_x86_Y libs into soc_x86 and arch_x86

Fixes: #8441

This commit merges the multiple sub libraries under soc_x86 and
arch_x86, e.g. arch_x86_Y_Z into a single arch_x86 libraries.

It also moves arch related stray files from libzephyr.a and places them
inside the arch_${ARCH} library, see #8826.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Merged n times arch_xtensa_Y libs into soc_xtensa and arch_xtensa

Fixes: #8441

This commit merges the multiple sub libraries under soc_xtense and
arch_xtensa, e.g. arch_xtensa_Y_Z into a single arch_xtensa libraries.

It also moves arch related stray files from libzephyr.a and places them
inside the arch_${ARCH} library, see #8826.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

SebastianBoe commented on CMakeLists.txt:
This is difficult to read.

Please put python code in python scripts.

tejlmand replied:
That can of course be done, but a small comment first.

Currently, cmake variables can be used directly inline, but if making a standalone script, those values must then be passed as arguments to the script.

With the size of the code, i.e. 6 lines, everything is visible directly here, no need to go to another file to see what is happening.

SebastianBoe replied:
Noted.

See https://github.com/tejlmand/zephyr/pull/1

SebastianBoe commented on CMakeLists.txt:
'kernel space symbols' implies the header file is encoded with symbols from the symbol table that are in kernel space. But in fact it is populated with:

```
#define NUM_KERNEL_OBJECT_FILES 4
#define KERNEL_OBJECT_FILE_0 zephyr/arch/libarch_arm.a
#define KERNEL_OBJECT_FILE_1 zephyr/subsys/bluetooth/libsubsys_bluetooth.a
#define KERNEL_OBJECT_FILE_2 zephyr/subsys/net/libsubsys_net.a
#define KERNEL_OBJECT_FILE_3 zephyr/kernel/libkernel.a

```

tejlmand replied:
will rephrase comment

tejlmand replied:
Done

tejlmand committed:
arch: x86: exception: Weak symbol on default exception handlers.

Exceptions can now be connected in x86 with weak symbols.

This ensures the linker will prefer the global defined exception over
the weakly defined version during linking.

The weakly defined handler is used in x86/core/fatal.c to ensure correct
selection of handler by the linker.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

pfalcon commented:
> Another nice part of the proposed structured, is that only the entry point to the library contains a CMakeLists.txt file, i.e. subsys/net/CMakeLists.txt and subsys/bluetooth/CMakeLists.txt
and other files are given more descriptive names,making navigation easier when having multiple *.cmake files open, instead of all files being named CMakeLists.txt

I would challenge that statement. Having all files named "CMakeLists.txt" allows everyone to know what's inside (including reusing knowledge from other projects seen), and search for such files/within such files easily (e.g. one can search for "foo" in CMakeLists.txt all subdirectories). Having files named inconsistently complicates all that.

And let me ask - is it possible to solve the issue this patch tries to solve (which seems to be removal for need of --whole-archive) without performing all those CMakeLists.txt -> *.cmake renamings? If it's anyhow possible, that's how it should be done, and renamings postponed until next PR (and separate discussion). Because this patch is massive at 226 changed files, which makes it very hard (== very long) to review, while majority of the changes are those renamings.

SebastianBoe commented:
> I would challenge that statement. Having all files named "CMakeLists.txt" allows everyone to know what's inside (including reusing knowledge from other projects seen), and search for such files/within such files easily (e.g. one can search for "foo" in CMakeLists.txt all subdirectories). Having files named inconsistently complicates all that.

I don't get this argument either. It is standard practice in many build systems to name the build scripts like this (Make, SCons, CMake, Meson). IMHO this argument should be omitted to keep the important issues in focus.

tejlmand commented:
Thanks for the comment.
Let me try to detail some more on this.

>>   Another nice part of the proposed structured, is that only the entry point to the library contains a
>> CMakeLists.txt file, i.e. subsys/net/CMakeLists.txt and subsys/bluetooth/CMakeLists.txt
>>   and other files are given more descriptive names,making navigation easier when having multiple  *.cmake files open, instead of all files being named CMakeLists.txt

> I would challenge that statement. Having all files named "CMakeLists.txt" allows everyone to know what's inside (including reusing knowledge from other projects seen), and search for such files/within such files easily (e.g. one can search for "foo" in CMakeLists.txt all subdirectories). Having files named inconsistently complicates all that.

I disagree that files are named inconsistently.
Files are still named *.cmake or CMakeLists.txt, where the CMakeLists.txt now has dedicated purpose.
This is what give a cleaner structure, imho.
One can always argue what is a clean structure, as that depends on each developers personal preference.

But by cleaner structure, I mean:
CMakeLists.txt contains an complete entity, i.e. it defines the libraries / targets that this unit provides.
This thus becomes the toplevel entry point, and Zephyr case there will be:
arch/CMakeLists.txt providing the arch library.
kernel/CMakeLists.txt providing kernel lib
subsys/<name>/CMakeLists.txt providing a subsys library
<app>/CMakeLists.txt: File containing an entry point when executing CMake for an app.

other *.cmake in respective subfolders are then used for a organizing the library.
Currently I have just re-used existing CMakeLists.txt files as content.

Searching for "foo" is still possible, just search *.cmake and CMakeLists.txt

> And let me ask - is it possible to solve the issue this patch tries to solve (which seems to be removal for need of --whole-archive) without performing all those CMakeLists.txt -> *.cmake renamings? If it's anyhow possible, that's how it should be done, and renamings postponed until next PR (and separate discussion). Because this patch is massive at 226 changed files, which makes it very hard (== very long) to review, while majority of the changes are those renamings.

Of course we can decide to do:
include(folder/CMakeLists.txt) instead of include(folder/file.cmake) 
only risks/limitation here are:

- a)  There can be multiple *.cmake files in same folder, but only a single CMakeLists.txt
This is not a problem today in Zephyr, but in previous projects I have worked on, it has been an advantage to be able to select the *.cmake file to use, and/or to split larger files into more, dedicated *.cmake files.
- b) Someone could accidentally do an add_subdirectory where an include should have been used.
add_subdirectory will fail if folder does not contain a CMakeLists.txt file
- c) Not getting the clean structure / separation between files as mentioned above

One reason why add_subdirectory cannot be used when reducing the number of static libraries because target_link_libraries(<target>) can only be used within the CMakeLists directory that defined the library.
But when using include(folder/file.cmake) you stay in same CMakeLists directory  and therefore the included *.cmake can issue a target_link_libraries(...)

I tried to keep the renaming in independent commits.
So that a commit would be: CMakeLists.txt -> *.cmake
or file content modifications.

But as you say.
The discussion on renaming can actually be taken separately.

tejlmand commented:
Has removed the renaming commit as proposed

tejlmand committed:
cmake: include file based on KConfig configuration

Issue: #8451

To support inclusion of CMakeLists.txt using the CMake function include
an extended version has been implemented which includes the file when
the corresponding KConfig setting is true.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Example showing the use of fewer libraries.

RFC for #8441

The RFC creates fewer libraries and uses CMake link libraries commands
to utilize CMake dependency handling
- drivers has been converted into a single library

It uses the populate source list presented in: #8442

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: creation of independent minimal libc library

Issue: #8441

To reduce number of libraries a libc library is created to which other
libs can link to (default linked through zephyr_interface).

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: cmake link libraries handling

Issues: #8441 #8851

As part of whole-archive removal it is important to specify correct
library dependencies and link ordering for the core libraries, i.e
kernel and arch libs.

To facilitate and handling of circular dependencies an
zephyr_app_linking INTERFACE library is constructed.
Any library linked to the zephyr_app_linking interface lib will be use
during final linking stage.

Using an interface library over a list has the advantage that libraries
can be added as dependency after the final link command is constructed
and hence remove any parsing ordering in CMake.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Removed app and zephyr libs from whole-archive

Issue: #8441

This commit remove the app and zephyr libraries from the
ZEPHYR_LIBS_PROPERTY.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand commented:
drivers, app and zephyr are now updated and removed from ${ZEPHYR_LIBS_PROPERTY} variable in CMake.

This also means that ${ZEPHYR_LIBS_PROPERTY} is empty in several samples / tests and for those.
Whenever ${ZEPHYR_LIBS_PROPERTY} is empty, the -Wl,--whole-archive flag is not used.

tejlmand committed:
cmake: Creating a single board library

RFC for #8441

The commit creates a single board libary for sources under the board
folder in Zephyr as part of #8441

It uses the populate source list presented in: #8442

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Creating a single openthread_platform library

RFC for #8441

The commit creates a single openthread_platform library for sources
under net/lib/openthread/platform folder in Zephyr as part of #8441

It uses the populate source list presented in: #8442

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Example showing the use of fewer libraries.

Issue #8441

This commits updates the following areas a creates specific libraries.
Libraries updated.
- subsys usb has been converted into a single library
- subsys mgmt has been converted into a single library
- subsys fs has been converted into a single library
- subsys app_memory has been converted into a single library
- ztest update to single lib and specifying linker flag for test_main

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Example showing the use of fewer libraries.

Issue #8441

This commits updates the following areas a creates specific libraries.
Libraries updated.
- lib/posix is a single library
- lib/json is still a normal cmake constructed library
- lib/cmsis_rtos is still a normal cmake constructed library

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

SebastianBoe commented:
> Do I understand correctly that, the –u<symbol1> option will make the linker pull in (in the 1st pass) the whole library (archive) where that symbol1 is defined like if it was "whole archived",

No, from the manual:

>Force symbol to be entered in the output file as an undefined symbol. Doing this may, for example, trigger linking of additional modules from standard libraries. -u may be repeated with different option arguments to enter additional undefined symbols. This option is equivalent to the "EXTERN" linker script command.

It just forces that symbol to be included. Not necessarily the whole archive, or the whole object file.
If that symbol is making references to other symbols then it might cause a chain-reaction of more symbols being included in the build. Just like any other symbol that the linker is trying to resolve.

aescolar commented:
@SebastianBoe : I think you missed my point:
> AND, it will not garbage collect any of those symbols’ sections later even if the original symbol1 is not used by anything

SebastianBoe commented:
> it will not garbage collect any of those symbols’ section later even if the original symbol1 is not used by anything?

Correct. symbol1 and it's dependencies will not be dropped.

aescolar commented:
My point was that, to the best of my understanding, now, when we use `-Whole-archive` together with `--gc-sections`, what we are doing is telling the linker: "in the 1st pass pick everything in that library/archive, if later you don't need it just drop it" == "we do not want to waste our time specifying library dependencies to provide the most optimal order of the libraries in the command line, just to save 20ms in link time if we are lucky to not have too circular dependencies. The linker knows what is used by what, so let it do it's job. It's not 1970 anymore, we can afford to waste a bit of compute time having the linker look thru a few unused symbols".
The -u option tells the linker "no matter what you (linker) think is needed, I want that symbol and its dependencies in the output".

Side comment: Note that a section is not the same as an archive. Very simple archives may only have 1 section. But that is not the case in general.
Note that what I referred to as the "linker 1st pass", is the linking step when the linker is noting down which symbols from the input it thinks are necessary, and at the same time is picking those it already sees as necessary. The garbage collection and the actual "linking itself" happen afterwards.

aescolar commented:
The corollary of all that being: If we use the -u option, then we indeed risk pulling unnecessary "stuff" into the binary/elf/whachawannacallit. While the current `-Whole-archive` together with `--gc-sections` should not.

SebastianBoe commented:
I am fairly certain that either method has the risk of including unneccessary "stuff" and the potential to produce the same desired optimal executable.

Today, KEEP in the LD script is used to express that although there are no references to a symbol, we are certain it is needed by something that the LD does not see (e.g. post build scripts).

In the new scheme -u (or actually, --require-defined, but that is a detail) will be used for the same effect. 

Due to misconfiguration (a bloated Kconfig presumably), either mechanism could bloat the application more than desired.

So I do not suspect we will see a code size hit, but we should for sure benchmark to make sure.

Note that this RFC is not about optimizing the LD runtime or the LD performance, but about resolving the issues in #8827. Which happens to require changing the linking mechanism.

tejlmand commented:
@aescolar 
The things you point out with the -u is absolutely correct, but let me try to explain the issue we are facing in Zephyr.

As a small example, let's take the hci_driver.c which is compiled into hci_driver.c.obj.
This source file contains only file scoped functions (static) and the _hci_driver_init function which is placed in the section .devconfig.init (as symbol: __config_sys_init__hci_driver_init2).

The linker script contains a KEEP for this section to avoid it being dropped.
When using whole-archive, then linker will consider all objects in the whole-archive during linking stage regardless if it consider them of need.
The KEEP flag will then ensures the .devconfig.init and its symbols not being dropped.

Now, removing the whole-archive, will still have the same effect on the linking process for objects taken into consideration by the linker.
So if linking the object file directly, it will keep the section as described in linker script.
But in our case, we place the object into an archive libA.a (containing: fileA.c.obj).
In this case the linker will consider fileA.c.obj only if it needs a symbol from the object file.
In the case with hci_driver.c.obj, the linker cannot see that it needs any symbols, and thus it will never consider to link this object file.

Now, in our case, we know that we need _hci_driver_init symbol to be kept, which we also specifies with the linker script. But object file is never taken into consideration cause there is not any symbol in the object file of interest to the linker.
By specifying -u_hci_driver_init we serves two purposes:
1) Tell the linker. Keep this symbol, cause we know it is in use
2) Ensure that the linker will actually take hci_driver.c.obj into consideration during linking, and hense the KEEP signal becomes of use

Now, in some libs, we might want to make use of the fact that linker should NOT consider an archive, such as with the kernel lib.

This is the reason for the suggestion of the:
zephyr_set_property(TARGET <lib> SYSINIT)
Telling the linker, this library contains symbols in SYSINIT (.devconfig.init) that you should always consider.
Libraries where we do not want this, we don't specify the property.

With the whole-archive approach, all libs created with zephyr_library will be whole-archived.
With the new approach, we have more fine grained control:
1) Not specify anything (Kernel). Linker takes care of everything.
2) Explicitly use -u<symbol>. Tell the linker keep this symbol and consider the object file during linking process
3) Set the property: zephyr_set_property(TARGET <lib> SYSINIT), Post-process archive after archiving and fetch symbols in the section named: .devconfig.init and tell linker to keep those

Note: Remember: .devconfig.init already has the KEEP flag
Note: SYSINIT is used, cause the macro used in C-code for specifying the symbol should be placed in section .devconfi.init is named: SYS_INIT() and is found in include/init.h

tejlmand commented:
@sebo 

> EDIT: For the record, I have previously claimed that removing --whole-archive will produce smaller executables, but I am not doing so any more.

We are not guaranteed that it will produce smaller libs, but there is a possibility.
Maybe not for Zephyr based libs, as those are very well controlled, but as described above, it provides a possibility to explicit tell if certain symbols should always be kept.
With the whole-archive on mostly everything, then anything placed in a keep section is ALWAYS kept, even if it potentially didn't have to.

aescolar commented:
@tejlmand Thanks Torsten,
Could you specify in a bit more detail what is the effect of setting
`zephyr_set_property(TARGET <lib> SYSINIT)`
Will this just add some option to the linker call related to this library? if so, which?

aescolar commented:
@tejlmand 
If I understand your thought process, this is:
a) Under the assumption that we want to get read of "whole-archive" all together.
b) We have a problem with all the symbols that we are picking in the linker script but otherwise seem unused for the linker in its 1st pass.
c) How do we solve that?

But, I really don't think there is any imperative need of stopping passing some libraries in the "whole-archive" part of the linker call. Therefore b) and c) are inconsequential.

It would be fine for me to not have some libraries in the "whole-archive" part of the linker call; if these were there only because of not having their dependencies to other libraries specified in cmake.

But, this `-u` approach would complicate things for developers, while effectively doing just the same otherwise.

tejlmand commented:
@aescolar 

> Could you specify in a bit more detail what is the effect of setting
zephyr_set_property(TARGET <lib> SYSINIT)
Will this just add some option to the linker call related to this library? if so, which?

Unfortunately the only option that exists to the linker is the -u option.
So what is done is a post process of the archive as part of the archiving  stages in CMake, where objdump is used to fetch the symbols in the .devconfig.init section (cause we know we always want to keep those).
That list of symbols are then used when creating the linker script so each symbol is placed with the keyword EXTERN in linker.cmd.

tejlmand commented:
@aescolar 

> a) Under the assumption that we want to get read of "whole-archive" all together.
b) We have a problem with all the symbols that we are picking in the linker script but otherwise seem unused for the linker in its 1st pass.
c) How do we solve that?

for a) there is already an issue: #6961 
and then of course b) and c) must be considered.

One can of course start questioning if #6961 is a real issue after all.
My original intention with this PR was to reduce the number of libraries, and while I was doing that, it made good sense to at the same time introduce a principle of getting rid of whole-archive, cause I'm mostly touching everything after all.

I already described the whole-archive itself above, so won't go through that again, but let me described one of the other things I see in relation to whole-archive.

In general, when we do software development, we strive to keep modules neat and with clear dependencies and architecture.
I think that is just as important i how we think of build systems and libraries.

Here CMake is very nice, cause it allows us to state those deps and CMake will do the link ordering.

With the whole-archive, we loose that part, cause everything is placed in a global property ${ZEPHYR_LIBS_PROPERTY} and everything added late in that process are lost, see: #8851 
The ${ZEPHYR_LIBS_PROPERTY} is needed to ensure the libs goes inside the whole-archive flags.

I made a small example:
https://raw.githubusercontent.com/tejlmand/cmake_playground/cmake_linking_example/CMakeLists.txt

In normal CMake we could specify:
`target_link_libraries(app1 a1)`
and with deps from a1 -> a2 -> a3
we would obtain:
`/usr/bin/cc   -rdynamic CMakeFiles/app1.dir/main.c.o  -o app1  liba1.a libb1.a libc1.a`

If you use CMake dependencies with whole archive, some libs will go outside:
`/usr/bin/cc   -rdynamic CMakeFiles/app2.dir/main.c.o  -o app2  -Wl,--whole-archive liba2.a -Wl,--no-whole-archive libb2.a libc2.a`

so instead everything is placed inside the whole-archive's flags, but then we looses the nice dependency handling:
`target_link_libraries(app3 -Wl,--whole-archive a3 b3 c3 -Wl,--no-whole-archive)`

With good CMake practice clearly see the dependency tree which I had in the presentation:
![ipsp_deps](https://user-images.githubusercontent.com/9802583/45222345-9b0d5600-b2b4-11e8-9279-5dbec19b72ae.png)

Compared to existing structure:
![ipsp_current](https://user-images.githubusercontent.com/9802583/45222459-f17a9480-b2b4-11e8-8640-655fb9076452.png)

tejlmand commented:
Another area to consider is 3rd party / customer's own libraries being linked in.

Those libraries may not be as configurable by nature, as Zephyr is.

A pre-built lib, may contain data in sections that are not desired to pull in, unless they are explicitly stated to be in use.

Maybe such libs should not be whole-archived, or maybe they should.
(And it may depend on the lib)

So it is still important to consider how the user of Zephyr can specify
a)	This lib should go into whole-archive / contains specific symbols to keep
b)	This lib should not be whole archived.

tejlmand committed:
cmake: cmake library handling

Issues: #8851

To support libraries outside zephyr tree the property
ZEPHYR_LIBS_PROPERTY has been removed as items added to this list is
lost in build process.

This commit moves the dependency relation setup to extensions.cmake.
It removes the empty library check as this is already a internal
feature of CMake.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

aescolar reviewed (changes requested):
Please split this PR in at least 2. 
Library reordering and setting dependencies between them is a quite different issue than the "whole-archive" one.
Independently of that.
I think the whole-archive solution you present here is just a lot of extra code & complexity to effectively end up with the same result.

tejlmand committed:
cmake: Placing subsys_bluetooth and subsys_net in kernel memory space

Issue: #8441

Setting memory space property on subsys_net and subsys_bluetooth
libraries.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Placing arch_${ARCH} in kernel memory space

Issue: #8441

Setting memory space property on arch_${ARCH} libraries.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Placing drivers in kernel memory space

Issue: #8441

Setting memory space property on drivers library.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
cmake: mbedtls interface

Added mbedtls interface library link dependency to zephyr_interface to
ensure correct propagation of mbedtls interface properties, such as
includes and config file.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

galak commented:
@tejlmand any plans to clean this up for 1.14?
